apps/src/dance/blocks.js

```diff
@@ -0,0 +1,188 @@
+import i18n from '@cdo/locale';
+
+// This color palette is limited to colors which have different hues, therefore
+// it should not contain different shades of the same color such as
+// ['#ff0000', '#cc0000', '#880000'].
+const limitedColours = [
+  // fully-saturated primary colors
+  '#ff0000', // RED
+  '#00ff00', // GREEN
+  '#0000ff', // BLUE
+
+  // fully-saturated secondary colors
+  '#ffff00', // YELLOW
+  '#00ffff', // CYAN
+  '#ff00ff', // MAGENTA
+
+  // some "tertiary" colors
+  '#ff8800', // ORANGE
+  '#8800ff', // PURPLE
+  '#00ff88' // LIME
+];
+
+const customInputTypes = {
+  spritePicker: {
+    addInput(blockly, block, inputConfig, currentInputRow) {
+      block.getVars = function() {
+        return {
+          [Blockly.BlockValueType.SPRITE]: [
+            block.getTitleValue(inputConfig.name)
+          ]
+        };
+      };
+      block.renameVar = function(oldName, newName) {
+        if (
+          Blockly.Names.equals(oldName, block.getTitleValue(inputConfig.name))
+        ) {
+          block.setTitleValue(newName, inputConfig.name);
+        }
+      };
+      block.removeVar = function(oldName) {
+        if (
+          Blockly.Names.equals(oldName, block.getTitleValue(inputConfig.name))
+        ) {
+          block.dispose(true, true);
+        }
+      };
+      block.superSetTitleValue = block.setTitleValue;
+      block.setTitleValue = function(newValue, name) {
+        if (
+          inputConfig.assignment &&
+          name === inputConfig.name &&
+          block.blockSpace.isFlyout
+        ) {
+          newValue = Blockly.Variables.generateUniqueName(newValue);
+        }
+        block.superSetTitleValue(newValue, name);
+      };
+
+      currentInputRow
+        .appendTitle(inputConfig.label)
+        .appendTitle(
+          new Blockly.FieldVariable(
+            null,
+            null,
+            null,
+            Blockly.BlockValueType.SPRITE,
+            i18n.sprite()
+          ),
+          inputConfig.name
+        );
+    },
+    generateCode(block, arg) {
+      return Blockly.JavaScript.translateVarName(block.getTitleValue(arg.name));
+    }
+  },
+  limitedColourPicker: {
+    addInput(blockly, block, inputConfig, currentInputRow) {
+      const options = {
+        colours: limitedColours,
+        columns: 3
+      };
+      currentInputRow
+        .appendTitle(inputConfig.label)
+        .appendTitle(
+          new Blockly.FieldColour('#ff0000', undefined, options),
+          'VAL'
+        );
+    },
+    generateCode(block, arg) {
+      return `'${block.getTitleValue(arg.name)}'`;
+    }
+  }
+};
+
+export default {
+  customInputTypes,
+  install(blockly, blockInstallOptions) {
+    // Legacy style block definitions :(
+    const generator = blockly.Generator.get('JavaScript');
+
+    const behaviorEditor = (Blockly.behaviorEditor = new Blockly.FunctionEditor(
```

Comment:
Worth removing some of these blocks that aren't used by Dance, or does Dance depend on behavior blocks?